src/TriangleNodesProvisioner.cpp

```diff
@@ -0,0 +1,199 @@
+        Fmask{ new index_vector_type((_NOrder+1)*NumFaces) },
+        Fx{ new real_matrix_type((_NOrder+1)*NumFaces, NumElements) },
+        Fscale{ new real_matrix_type((_NOrder+1)*NumFaces, NumElements) },
+        EToV{ new index_matrix_type(NumElements, NumFaces) },
+        EToE{ new index_matrix_type(NumElements, NumFaces) },
+        EToF{ new index_matrix_type(NumElements, NumFaces) },
+        vmapM{ new index_vector_type((_NOrder+1)*NumFaces*NumElements) },
+        vmapP{ new index_vector_type((_NOrder+1)*NumFaces*NumElements) },
+        Mesh2D { _MeshManager },
+        Nodes1D{ new Nodes1DProvisioner(_NOrder, NumElements, -1.0, 1.0) },
+		Jacobi{}, Vandermonde{}, LinSolver{}
+    {}
+
+
+	void TriangleNodesProvisioner::evaluateSimplexPolynomial(const real_vector_type & a, const real_vector_type & b, const index_type i, const index_type j, real_vector_type & p) const {
+		real_vector_type h1(a.length(0));
+		real_vector_type h2(b.length(0));
+
+		Jacobi.computeJacobiPolynomial(a, 0.0, 0.0, i, h1);
+		Jacobi.computeJacobiPolynomial(b, 2.0*i+1.0, 0.0, j, h2);
+
+		real_vector_type c(b.length(0));
+
+		p = sqrt(2.0)*h1*h2*pow(1.-b, i);
+	}
+
+    void TriangleNodesProvisioner::rsToab(const real_vector_type & r, const real_vector_type & s, real_vector_type & a, real_vector_type & b) const {
+        int Np = r.length(0);
+        for( index_type i=0; i < Np; i++) {
+            if(s(i) != 1.0) 
+                a(i) = 2.0*(1.0+r(i))/(1.0-s(i)) - 1.0;
+            else
+                a(i) = -1.0;
+        }
+        b = s;
+    }
+
+    void TriangleNodesProvisioner::xyTors(const real_vector_type & x, const real_vector_type & y, real_vector_type & r, real_vector_type & s) const {
+        const index_type Np = x.length(0);
+        real_vector_type L1(Np), L2(Np), L3(Np);
+        L1 = (sqrt(3.0)*y+1.0)/3.0;
+        L2 = (-3.0*x - sqrt(3.0)*y + 2.0)/6.0;
+        L3 = ( 3.0*x - sqrt(3.0)*y + 2.0)/6.0;
+
+        r =-L2 + L3 - L1;
+        s =-L2 - L3 + L1;
+    }
+
+
+    void TriangleNodesProvisioner::computeVandermondeMatrix(const int N, const real_vector_type & r, const real_vector_type & s, const real_matrix_type & V) const {
+        const index_type Nr = r.length(0);
+
+        real_vector_type a(Nr), b(Nr);
+        rsToab(r, s, a, b);
+
+        // build the Vandermonde matrix
+        index_type count = 0;
+        for (index_type i=0; i < N + 1; ++i) {
+            for (index_type j=0; j < N - i + 1; ++j) {
+                real_vector_type p(Nr);
+                evaluateSimplexPolynomial(a,b,i,j,p);
+                V(Range::all(),count) = p;
+                ++count;
+            }
+        }
+    }
+
+    void TriangleNodesProvisioner::computeEquilateralNodes(real_vector_type & x, real_vector_type & y) const {
+        real_vector_type alphaOptimal(15);
+
+        alphaOptimal = 0.0000,0.0000,1.4152,0.1001,0.2751,0.9800,1.0999,
+                       1.2832,1.3648,1.4773,1.4959,1.5743, 1.5770,1.6223,1.6258;
+
```

Comment:
I would define alphaOptimal as a static const array:
`static const real_vector_type alphaOptimal[16] = {0.0000,0.0000,1.4152,0.1001,0.2751,0.9800,1.0999,1.2832,1.3648,1.4773,1.4959,1.5743, 1.5770,1.6223,1.6258};`

Reply:
Is there a closed form for those values? Does it affect anything that they are rounded to 4 places?

Reply:
The values slightly change the location of nodes on the triangle. They appear to converge rapidly to a closed-form value of 5/3 at around NOrder=15/16,  but it doesn't not look like there are closed form values for smaller orders. See here: https://github.com/dsteinmo/euler-nullproj/blob/master/NUDG/Codes2D/Nodes2D.m#L6

I think the point is that 4 digits of accuracy is sufficient enough such that interpolation does not have any of the wild oscillations that are associated with the equally-distributed nodes.